[Android] Add Standalone Note Feature

hemanggs opened the pull request:
Related issues :
https://github.com/organicmaps/organicmaps/issues/266
https://github.com/organicmaps/organicmaps/issues/7620
https://github.com/organicmaps/organicmaps/issues/9003

Addresses 
- Currently when an OSM note is created it is placed automatically to the center of area. It is not convenient if you want to add a note, for example, in some part of the forest (and it is put in the center).

- When user is unable to find a POI type for the feature they want to add. At that point, they should be given the option to create a note instead of a POI , and later use advanced editors to make additions/edits.

- For Places with atleast one tag ,i am  using the existing Create note API , for places without  a single  tag using `Editor::CreateStandaloneNote(ms::LatLon const & latLon, std::string const & noteText)` , as the existing API requires Editable map object which is not loaded in case of POI without atleast one Tag .

https://github.com/user-attachments/assets/ab843a08-ba39-4eaf-a77f-aa5723e516b4

![normal button wala ss](https://github.com/user-attachments/assets/f993d8dd-9755-44f7-b759-58eca50fa21c)



https://github.com/user-attachments/assets/3a50f8aa-16b5-4857-8140-196f24ac52a8

![cant find category wala note](https://github.com/user-attachments/assets/52e74112-197e-4a36-a4ca-c166c0748088)



 - Todos : 
	- [ ] Better placement for the button in Place Page ( For places with Alteast one tag , i see 3 buttons , not good UI  need a better way)
	- [ ] Better add note text dialog box UI?
	- [ ] Maybe have option to toggle this in settings ? 
	- [ ] Test coverage ? 

- Future Tasks : 
	- [ ] Share/publish existing bookmarks with names + descriptions as OpenStreetMap.org notes (and then track their status/responses/chats)
	- [ ] Combine Add Place/Edit Place and make them work smart, depending on the context
	
@AndrewShkrob  @biodranik

AndrewShkrob commented:
How about this UI:

User not logged in OSM -> show only one button - Edit Place
Click `Edit Place` -> login -> go back to the place page -> show the same section as for donations
<img width="539" alt="image" src="https://github.com/user-attachments/assets/35e4fda6-e59d-414a-9788-16524ac17c10" />

<table>
  <thead>
    <tr>
      <th colspan="3" align="left"><strong>OpenStreetMap</strong></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td align="center">Add Place</td>
      <td align="center">Edit Place</td>
      <td align="center">(Remove Place)</td>
    </tr>
    <tr>
      <td colspan="3" align="center">Leave note to OpenStreetMap</td>
    </tr>
  </tbody>
</table>


By clicking on the "Leave note" button it will transform into something like this:

<table>
  <thead>
    <tr>
      <th colspan="3" align="left"><strong>OpenStreetMap</strong></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td align="center">Add Place</td>
      <td align="center">Edit Place</td>
      <td align="center">(Remove Place)</td>
    </tr>
    <tr>
      <td colspan="3" align="center">A huge EditView</td>
    </tr>
    <tr>
      <td colspan="1" align="center">Cancel</td>
      <td colspan="2" align="center">Publish Note</td>
    </tr>
  </tbody>
</table>

With this UI 
* we will reduce the amount of buttons displayed for the users who are not contributing to OSM
* we won't need to show the awkward dialog for the note input
* we will follow the style we already have - the "cards" for descriptions and donations

hemanggs commented:
Yeah , this UI is way better , thanks.

- For user not logged in  , Variant 2 along with buttons that redirect to existing osm login activity shows even lesser buttons , seems better ? 

- User Logged in & note input UI looks good, for the third button "Remove Place"  easy option is to reuse , "Place does not exist" button , or this opportunity can be used to discuss implementation of actual POI deletion (https://github.com/organicmaps/organicmaps/issues/1650) or adding the lifecycle prefix disused (eg. disused:amenity=cafe we can allow to Mark POI as disused when in editing mode. Ref : https://github.com/organicmaps/organicmaps/issues/9542 ) can be discussed if required . 

-  After use clicks selects "Add/Edit Place" button  , we display all the same options that are currently shown as another buttons screen/dialog/drop down? , should we stick to existing buttons or is it better to combine "add business" with "add place" and "edit business" with "edit place" , what is the use case of having them separate? 

- Is it required to address this scenario where  Many cases in OM show multiple buttons, when selected buildings we get edit place and Add business (i'm guessing as buildings can have shops within while buildings can have own tags as whole too)  selecting parks/scrub/orchard similar  shows both add place and edit place and other cases where multiple options are currently shown within OM flow

biodranik commented:
- Two or more OSM buttons on the place page may be inconvenient for non-OSMers
- I do not understand the use case for adding a note on a failed search. Notes have coordinates. If the user searched and didn't find anything, then the place was not found anywhere, not just at the screen center where a note is created.
- Edit place is displayed on non-editable farmlands etc., only to show two useful buttons inside (leave a note and mark place as not existing). That's not "editing".
- Leaving OSM note is a feature for a small number of expert users, and it may be confusing for those who don't know what it is about (is it a note for myself like a bookmark note, but on a map?)
- The idea of showing more UI for logged users is good, although we also want anyone to contribute if it is possible and doesn't generate much spam

hemanggs commented:
> * Two or more OSM buttons on the place page may be inconvenient for non-OSMers

Totally agree , Ideas suggested by Andrei would occupy space of two buttons i feel , is there a better UX case for a single button like "Contribute to OSM"  which has drop down/pop up/ another screen which has more option? 

 I feel Edit place on Place page makes sense to have separate as having edit place not on place page may be confusing for users as supposed to having it on selected place's place page .  

Two buttons will be required either case i feel ? 

> * I do not understand the use case for adding a note on a failed search. Notes have coordinates. If the user searched and didn't find anything, then the place was not found anywhere, not just at the screen center where a note is created.

I think theres some mis understanding , the missing search is when user is adding a POI , and from issues i saw in the past users complained about OM not supporting some features they wanna add , so they instead prefer leaving a note and using other editors / OSM website to make changes later , that was the intended use case 
(Ref : https://github.com/organicmaps/organicmaps/issues/7620#issuecomment-2002624272 , https://github.com/organicmaps/organicmaps/issues/266#issuecomment-902528157 , https://github.com/organicmaps/organicmaps/issues/7620 , https://github.com/organicmaps/organicmaps/issues/9003#issue-2471868887) 


> * Edit place is displayed on non-editable farmlands etc., only to show two useful buttons inside (leave a note and mark place as not existing). That's not "editing".

Yeah this is exactly my point , is it better to not display the edit button for these cases ( i know theres multiple scrub, lake , orchard ... etc )  

> * Leaving OSM note is a feature for a small number of expert users, and it may be confusing for those who don't know what it is about (is it a note for myself like a bookmark note, but on a map?)

I think better naming can handle this ? something like this  "Send a note to the OpenStreetMap editors" , technically thats how this feature is meant to be used i think ?  people leaving notes to themselves ( osm editors them selves ) is a good additional case.

biodranik commented:
1. Agree, let's discuss the "Leave a Note" button in the Place Page as a part of its redesign.
2. Now implementing a note only when added feature type was not found in the OSM editor is already a great improvement.
3. Would it be possible to integrate EditText and Send buttons under the already displayed "Can't find a suitable category?" messages, to avoid unnecessary click, and keep the typed text there until the app is restarted/the send button is pressed?

hemanggs commented:
> 3. Would it be possible to integrate EditText and Send buttons under the already displayed "Can't find a suitable category?" messages, to avoid unnecessary click, and keep the typed text there until the app is restarted/the send button is pressed?

Agreed that seems better, i'll implement it and share how it looks

biodranik commented on android/app/src/main/res/values/strings.xml:
```diff
@@ -602,6 +602,9 @@
     <string name="error_enter_correct_vk_page">Enter a valid VK username or web address</string>
     <string name="error_enter_correct_line_page">Enter a valid LINE ID or web address</string>
     <string name="placepage_add_place_button">Add Place to OpenStreetMap</string>
+    <string name="leave_osm_note">Leave Note to OpenStreetMap</string>
```
Maybe reuse `editor_other_info`?

hemanggs replied:
`<string name="editor_other_info">Note to OpenStreetMap volunteers (optional)</string>`
 
 Should we have the "(Optional)" on the button ?

biodranik replied:
After some testing, a simpler "Send" button would be enough `editor_report_problem_send_button`

It feels like a toast after pressing send/closing the editor UI is also needed, WDYT?

biodranik reviewed (commented):
Thanks! Looks good. Is there a layout issue in landscape? (Android 5)

![grafik](https://github.com/user-attachments/assets/788d093e-1f5a-4f6a-a158-a3983ae33679)

hemanggs commented on editor/osm_editor.cpp:
```diff
@@ -1187,6 +1187,12 @@ void Editor::CreateNote(ms::LatLon const & latLon, FeatureID const & fid,
+void Editor::CreateStandaloneNote(ms::LatLon const & latLon, std::string const & noteText)
+{
+  CHECK_THREAD_CHECKER(MainThreadChecker,(""));
+  m_notes->CreateNote(latLon, noteText + "\n");
```
its to separate the note text from `#organicmaps android` footer

![image](https://github.com/user-attachments/assets/3743832e-83e1-41b5-adc8-704976a00e72)

hemanggs commented on android/app/src/main/res/values/strings.xml:
```diff
@@ -602,6 +602,9 @@
     <string name="error_enter_correct_line_page">Enter a valid LINE ID or web address</string>
     <string name="placepage_add_place_button">Add Place to OpenStreetMap</string>
+    <string name="leave_osm_note">Leave Note to OpenStreetMap</string>
+    <string name="osm_note_hint">Describe the issue, add details, or suggest an improvement.</string>
```
`    <string name="editor_note_hint">Describe errors on the map or things that cannot be edited with Organic Maps</string>
`
Add place button leading to "Describe errors" , could this be confusing for the user.  Also It has `edited`  when user chose add place button.

my hint is not optimal too i feel, considering its the category select screen  

Is something like this better since it is in the category select screen
 "Couldn't find what you are looking for , leave a note to OpenStreetMap volunteers?"

biodranik replied:
Agree that it's confusing a bit. After some testing, and taking into an account the long existing text message above (about missing category in OM), a hint something like this may look clearer:

> Organic Maps allows to add simple point categories only, that means no towns, roads, lakes, building outlines, etc. Please add such categories directly to <a href="https://www.openstreetmap.org">OpenStreetMap.org</a>. Check our <a href="https://organicmaps.app/faq/editing/advanced-map-editing">guide</a> for detailed step by step instructions.

`Or, alternatively, leave a note to OpenStreetMap community so someone else can add or fix a place here.`

hemanggs commented on android/app/src/main/res/values/strings.xml:
```diff
@@ -602,6 +602,9 @@
     <string name="placepage_add_place_button">Add Place to OpenStreetMap</string>
+    <string name="leave_osm_note">Leave Note to OpenStreetMap</string>
+    <string name="osm_note_hint">Describe the issue, add details, or suggest an improvement.</string>
+    <string name="error_enter_note">Please enter a note text.</string>
```
Yeah , agreed the toast is not good UX , i'll fix this

hemanggs commented:
PTAL 


https://github.com/user-attachments/assets/f90650c9-c394-4f32-bfd7-14510b7a1700

biodranik commented on android/app/src/main/java/app/organicmaps/sdk/editor/Editor.java:
```diff
@@ -162,6 +162,7 @@ public static void createMapObject(FeatureCategory category)
   public static native void nativeCreateNote(String text);
```
Can this implementation be reused for the standalone note? Isn't mapObject containing the same lat lon?

hemanggs replied:
> For Places with atleast one tag ,i am using the existing Create note API , for places without a single tag using Editor::CreateStandaloneNote(ms::LatLon const & latLon, std::string const & noteText) , as the existing API requires Editable map object which is not loaded in case of POI without atleast one Tag .

Yes i did think of reusing that , but the issue is that it uses `g_editableMapObject` (see Editor.cpp)  , and the data is updated/loaded into `g_editableMapObject` after a category is chosen  from category screen  , since our note is being created in category selection screen , the note instead gets created at  the previously to `g_editableMapObject`  POI , instead of the selected location. `nativeCreateNote` can only be used when `g_editableMapObject` has been loaded with a category , which defeats the purpose here. 

Squashed

biodranik replied:
Should g_editableMapObject be initialized when crosshair selection on the map is done?

hemanggs replied:
I dont think `g_editableMapObject` is allowed to be initialized without a type currently , but can be implemented to initialize with the lat , lon and without a type as soon as crosshair is selected and add type after category is selected , is it better to do it that way?

hemanggs commented:
i think `fragment_report.xml`  &  `ReportFragment.java` have been deprecated ?

biodranik commented:
> Just that i couldn't find it being used in the android app anywhere.

ReportFragment class is used in the code.

biodranik commented on android/app/src/main/java/app/organicmaps/editor/FeatureCategoryFragment.java:
```diff
@@ -104,4 +115,59 @@ public void selectCategory(FeatureCategory category)
+  @Override
+  public void onSendNoteClicked()
+  {
+    if (!OsmOAuth.isAuthorized(requireContext()))
```
```suggestion
    if (!OsmOAuth.isAuthorized())
```

biodranik reviewed (commented):
Thanks! Please rebase, it is almost ready to be merged )

hemanggs commented:
Sorry for the delay.

- Rebased.

> It feels like a toast after pressing send/closing the editor UI is also needed, WDYT?

- Added a toast PTAL(will rebase before merge)

biodranik commented on android/app/src/main/java/app/organicmaps/editor/FeatureCategoryFragment.java:
```diff
@@ -104,4 +115,61 @@ public void selectCategory(FeatureCategory category)
+    }
+  }
+
+  private void showNoteConfirmationDialog(double lat, double lon, String noteText)
```
nit: add a comment that it duplicates `showNoobDialog()`

biodranik commented on android/app/src/main/res/values/strings.xml:
```diff
@@ -602,6 +602,9 @@
     <string name="placepage_add_place_button">Add Place to OpenStreetMap</string>
+    <string name="osm_note_hint">Or, alternatively, leave a note to OpenStreetMap community so that someone else can add or fix a place here.</string>
+    <string name="osm_note_toast">Note sent to OpenStreetMap</string>
+    <string name="error_enter_note">Please enter a note text.</string>
```
Remove?

biodranik commented on android/app/src/main/res/values/strings.xml:
```diff
@@ -602,6 +602,9 @@
     <string name="error_enter_correct_line_page">Enter a valid LINE ID or web address</string>
     <string name="placepage_add_place_button">Add Place to OpenStreetMap</string>
+    <string name="osm_note_hint">Or, alternatively, leave a note to OpenStreetMap community so that someone else can add or fix a place here.</string>
+    <string name="osm_note_toast">Note sent to OpenStreetMap</string>
```
AFAIR it is not sent immediately:
```suggestion
    <string name="osm_note_toast">Note will be sent to OpenStreetMap</string>
```

biodranik reviewed (approved):
Thanks! LGTM with a few minor comments.

hemanggs commented:
- Applied clang format
- fixed nits

PTAL
